## NYC Taxi Fare Prediction using Eexploratory Data Analysis and Machine Learning

Source Dataset: https://data.cityofnewyork.us/Transportation/2017-Yellow-Taxi-Trip-Data/biws-g3hs/about_data  
Data Provided By: Taxi and Limousine Commission (TLC)  
Dataset Owner: NYC OpenData

#### Import libraries

In [5]:
# Basic imports for data analysis
import numpy as np
import pandas as pd
import datetime as dt

# Importing libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Importing libraries for data modeling
from sklearn.model_selection import train_test_split


#### Review Data Available

In [ ]:
# Load dataset
df_original= pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
df_original.head()

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80

In [4]:
# Variable overview
df_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

**Observations**:
- Compared to the original dataset, which contains 113M records, we are working with a smaller subset of the data with 22,699 records as part of practice.  
- The column 'Unnamed: 0' has not been mentioned in the data dictionary and is likely to random no. generated as index.
- Columns 'tpep_pickup_datetime' and 'tpep_dropoff_datetime' are date/time variables, which would need to be converted to the correct format.

In [8]:
# Summary statistics
df_original.describe(include= 'all')

Unnamed: 0      VendorID   tpep_pickup_datetime  \
count   2.269900e+04  22699.000000                  22699   
unique           NaN           NaN                  22687   
top              NaN           NaN  04/15/2017 6:05:19 PM   
freq             NaN           NaN                      2   
mean    5.675849e+07      1.556236                    NaN   
std     3.274493e+07      0.496838                    NaN   
min     1.212700e+04      1.000000                    NaN   
25%     2.852056e+07      1.000000                    NaN   
50%     5.673150e+07      2.000000                    NaN   
75%     8.537452e+07      2.000000                    NaN   
max     1.134863e+08      2.000000                    NaN   

        tpep_dropoff_datetime  passenger_count  trip_distance    RatecodeID  \
count                   22699     22699.000000   22699.000000  22699.000000   
unique                  22688              NaN            NaN           NaN   
top     10/18/2017 8:07:45 PM              NaN            NaN           NaN   
freq                        2              NaN            NaN           NaN   
mean                      NaN         1.642319       2.913313      1.043394   
std                       NaN         1.285231       3.653171      0.708391   
min                       NaN         0.000000       0.000000      1.000000   
25%                       NaN         1.000000       0.990000      1.000000   
50%                       NaN         1.000000       1.610000      1.000000   
75%                       NaN         2.000000       3.060000      1.000000   
max                       NaN         6.000000      33.960000     99.000000   

       store_and_fwd_flag  PULocationID  DOLocationID  payment_type  \
count               22699  22699.000000  22699.000000  22699.000000   
unique                  2           NaN           NaN           NaN   
top                     N           NaN           NaN           NaN   
freq                22600           NaN           NaN           NaN   
mean                  NaN    162.412353    161.527997      1.336887   
std                   NaN     66.633373     70.139691      0.496211   
min                   NaN      1.000000      1.000000      1.000000   
25%                   NaN    114.000000    112.000000      1.000000   
50%                   NaN    162.000000    162.000000      1.000000   
75%                   NaN    233.000000    233.000000      2.000000   
max                   NaN    265.000000    265.000000      4.000000   

         fare_amount         extra       mta_tax    tip_amount  tolls_amount  \
count   22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean       13.026629      0.333275      0.497445      1.835781      0.312542   
std        13.243791      0.463097      0.039465      2.800626      1.399212   
min      -120.000000     -1.000000     -0.500000      0.000000      0.000000   
25%         6.500000      0.000000      0.500000      0.000000      0.000000   
50%         9.500000      0.000000      0.500000      1.350000      0.000000   
75%        14.500000      0.500000      0.500000      2.450000      0.000000   
max       999.990000      4.500000      0.500000    200.000000     19.100000   

        improvement_surcharge  total_amount  
count            22699.000000  22699.000000  
unique                    NaN           NaN  
top                       NaN           NaN  
freq                      NaN           NaN  
mean                 0.299551     16.310502  
std                  0.015673     16.097295  
min                 -0.300000   -120.300000  
25%                  0.300000      8.750000  
50%                  0.300000     11.800000  
75%                  0.300000     17.800000  
max     

**Observations**:
- There seems to be outlying values for fare_amount, trip_distance, total_amount etc. as the max. value of ~1000 USD is much higher than the mean.
- We have negative values in variables fare_amount, exttra etc.

Based on our observations from above, we can plan our EDA approach as below:
- Data Cleaning
    - Check/eliminate/deal with duplicates
    - Check/eliminate/deal with Nulls
    - Modify columns as needed
    - Identify outliers, isolate to assess how to deal with them
    - Check for incorrect values

- Draw insights from variables
    - Derive 'trip_duration' column based on date/time values provided
    - Can trip distance and trip duration be correlated so we keep just one?
    - Does the data sample capture all kind of rate code IDs?
    - Is there any variance in nos. observed between different data providers?

#### Exploratory Data Analysis

#### Model Selection/Checking Assumptions

#### Model Build and Initial Assessment

#### Final Assessment and Summary